File: Python/heap_sort.py
def heap_sort(arr):
    for i in range(len(arr) // 2, -1, -1):
        heapify(arr, len(arr), i)

    for i in range(len(arr) - 1, -1, -1):
        arr[0], arr[i] = arr[i], arr[0]
        heapify(arr, i, 0)


def heapify(arr, len, root):
    if root >= len:
        return
    left = 2 * root + 1
    right = 2 * root + 2

    largest = root
    if left < len and arr[left] > arr[largest]:
        largest = left
    if right < len and arr[right] > arr[largest]:
        largest = right

    if largest != root:
        arr[root], arr[largest] = arr[largest], arr[root]
        heapify(arr, len, largest)


if __name__ == "__main__":
    arr = [4, 3, 2, 5, 10, 2]
    heap_sort(arr)
    print(arr)

File: Java/HeapSort.java
package Java;

import java.util.Arrays;

class HeapSort {
    public void sort(int[] arr) {
        if (arr == null || arr.length == 0) {
            return;
        }

        for (int i = arr.length / 2; i >= 0; i--) {
            heapify(arr, arr.length - 1, i);
        }

        for (int i = arr.length - 1; i > 0; i--) {
            swap(arr, 0, i);
            heapify(arr, i - 1, 0);
        }
    }

    private void heapify(int[] arr, int end, int root) {
        if (root > end) {
            return;
        }
        int largest = root;
        int left = root * 2 + 1;
        int right = root * 2 + 2;
        if (left <= end && arr[left] > arr[largest]) {
            largest = left;
        }
        if (right <= end && arr[right] > arr[largest]) {
            largest = right;
        }

        if (largest != root) {
            swap(arr, largest, root);
            heapify(arr, end, largest);
        }
    }

    private void swap(int[] arr, int index1, int index2) {
        int tmp = arr[index1];
        arr[index1] = arr[index2];
        arr[index2] = tmp;
    }

    public static void main(String args[]) {
        HeapSort hs = new HeapSort();
        int[] arr = { 5, 2, 8, 3, 6, 1 };
        hs.sort(arr);

        System.out.println("Array after sort is: " + Arrays.toString(arr));
    }
}
